java/JavaParser.g4

```diff
@@ -463,13 +463,14 @@ expression
     : primary
     | expression bop='.'
       (IDENTIFIER
+      | IDENTIFIER '(' expressionList? ')'
       | THIS
       | NEW nonWildcardTypeArguments? innerCreator
       | SUPER superSuffix
       | explicitGenericInvocation
       )
     | expression '[' expression ']'
-    | expression '(' expressionList? ')'
+    | IDENTIFIER '(' expressionList? ')'
```

Comment:
I disagree with your solution. Other code analysis tools have the same invocation expressions. Take a look at [InvocationExpressionSyntax](http://www.coderesx.com/roslyn/html/49CEBC83.htm) in Roslyn: it contains `ExpressionSyntax` that represents the expression part of the invocation and `ArgumentList` that represents the list of arguments of the invocation. So, there are no big problems of such invocations.

You can use try to add [alternative labels](https://github.com/antlr/antlr4/blob/master/doc/parser-rules.md#alternative-labels) for more convenient and short Visitor methods.

Reply:
Sorry, I don't get your point? And I can't see how a label would help here.

My problem with the current solution is that using the AST to identify a method call is unnecessary hard. When visiting an expression, e.g. `System.out.println("hello world!");` the left branch contains something that looks like a field access for a field called `println`, and to detect that it is a method call you have to walk many levels up again (where the number of levels is proportional to the length of the part before the bracket). Or you have to examine the right branch first and then pass on the information 'last expression is a method call' when traversing the left branch.

With the proposed change everything is available in just one node.

Is your comment for both changed lines? The rule you annotated is just for 'local' method calls, it could be combined with the rule which is spread over line 464 and 466 like this:

     (expression bop='.')? IDENTIFIER '(' expressionList? ')'

I just wanted to keep the number of changes low.

If there are real issues with the change I'm happy to rework the pull request. If you are saying "_Others are doing it the same way_": This sounds like bureaucracy, not like software development ;)

Reply:
You have a couple of member reference expression on the left side:

```antlr
expression bop='.'
      (IDENTIFIER
      | THIS
      | NEW nonWildcardTypeArguments? innerCreator
      | SUPER superSuffix
      | explicitGenericInvocation
      )
```

and one invocation expression that wraps member reference and `expressionList`:

```antlr
    | expression '(' expressionList? ')'
```

The final parse tree looks like this:

```lisp
(expression
  (expression
    (expression
      (expression
        (primary
          'System'
        )
      )
      '.'
      'out'
    )
    '.'
    'println'
  )
  '('
  (expressionList
    (expression
      (primary
        (literal
          '"hello world!"'
        )
      )
    )
  )
  ')'
)
```

You can detect invocation expression by open and close parenthesis. I don't see access troubles.

Reply:
Yes, you can, it`s just software after all ;). But it could be so much easier if the parse tree looks like this:

        (expression
          (expression
            (expression
              (primary
                'System'
              )
            )
            '.'
            'out'
          )
          '.'
          (expression
            'println'
            '('
              (expressionList
                (expression
                  (primary
                    (literal
                      '"hello world!"'
                    )
                  )
                )
              )
            ')'
          )
        )

I don't really see an argument what speaks against the change. A cite from [Wikipedia "Left recursion pitfalls"](https://en.wikipedia.org/wiki/Left_recursion#Pitfalls): _[..] but less faithful to the input and much harder to implement for some operators [..]_.

Reply:
If you are afraid that performance is an issue, a quick check showed that it is not. In a quick test this branch is even a bit faster. Not noteworthy, just 71 vs 73 seconds for the existing test files with 100 iterations. For a special test file (10.000 method calls) the branch is 5 seconds faster with 74 vs 79 seconds, again for 100 iterations.

Reply:
I don't like two duplicated rules in your change. But if you insist I'm forced to refer to @parrt and other contributors for opinion.

Reply:
Ah sorry, I didn't get it. If it's just the duplication I could change my pull request. I did it this way because I wanted to keep the number of changes in the grammar minimal. 

So what about extracting a new rule? E.g. like this:

    methodCall:
        : IDENTIFIER '(' expressionList? ')'
        ;

This could be than used twice:

    expression
        : primary
        | methodCall
        | expression bop='.'
          (IDENTIFIER
          | methodCall
          | THIS
          | NEW nonWildcardTypeArguments? innerCreator
          | SUPER superSuffix
          | explicitGenericInvocation
          )

Would this fit better?

Reply:
Honestly, I don't think so. I think the current solution is good.

Reply:
I did a force push to my clone. I kept both changes as separate commits to keep the history for review. I will squash them if this general change is now acceptable.

The refactoring also fixed issue #1040 (I didn't investigated why) but decreases the performance gain a bit. But performance was not a goal of this change, so as long as the performance does not get worse it should be acceptable.

Reply:
Moreover, the new change in `expression` can break old code that relies on grammar that exists for very long time.